src/server/routes/search.rs

```diff
@@ -123,11 +123,34 @@ async fn results(
         config.safe_search,
     );
 
-    let cache_key = format!(
+    let mut cache_key = format!(
         "http://{}:{}/search?q={}&page={}&safesearch={}",
         config.binding_ip, config.port, query, page, safe_search_level
     );
 
+    // Modify the cache key adding each enabled search engine to the string
+    if let Some(cookie_value) = &cookie_value {
+        let mut engines: Vec<String> = cookie_value
+            .engines
+            .iter()
+            .map(|s| String::from(*s))
+            .collect::<Vec<String>>();
+
+        // We sort the list of engine so the cache keys will match between users. The cookie's list of engines is unordered.
+        engines.sort();
+        cache_key = cache_key + &(engines.join(""));
+    } else {
+        let mut engines: Vec<String> = config
+            .upstream_search_engines
+            .iter()
+            .filter(|map| *map.1)
+            .map(|map| String::from(map.0))
+            .collect();
+
+        engines.sort();
+        cache_key = cache_key + &(engines.join(""));
+    }
```

Comment:
```suggestion
    // Modify the cache key adding each enabled search engine to the string
    if let Some(cookie_value) = &cookie_value {
        let mut engines: Vec<String> = cookie_value
            .engines
            .iter()
            .map(|s| String::from(*s))
            .collect::<Vec<String>>();

        // We sort the list of engine so the cache keys will match between users. The cookie's list of engines is unordered.
        engines.sort();
        cache_key = cache_key + &(engines.join(""));
    } else {
        let mut engines: Vec<String> = config
            .upstream_search_engines
            .iter()
            .filter(|map| *map.1)
            .map(|map| String::from(map.0))
            .collect();

        engines.sort();
        cache_key = cache_key + &(engines.join(""));
    }
```

I think I would suggest instead of having a duplicate cookie handling code, you can add the search engine addition code for the URL search parameter in the already existing cookie handling code which can be found here:

https://github.com/neon-mmd/websurfx/blob/5b4864424a3f740f13749185505385ed66f46644/src/server/routes/search.rs#L156-L212  

Also, instead of adding the search engines with ("") I would suggest adding them comma separated (",") which I think would be better.

Moreover, instead of using (+) for concatenation, I would recommend using the `format!` macro, which tends to be more idiomatic and rusty. :slightly_smiling_face:

Reply:
I can fix the separator and concatenation but I'm confused by what you mean for the cookie handling code? Do you want me to move the use of some of the cookie's variables `engines` to before the cache key is built and modified and then reuse it later on when it was used before?

Reply:
> I can fix the separator and concatenation but I'm confused by what you mean for the cookie handling code? Do you want me to move the use of some of the cookie's variables `engines` to before the cache key is built and modified and then reuse it later on when it was used before?

No I am actually suggesting to move the code related to `engines` and everything within the already existing cookie handling code present in the `results` function. I hope you got my point :slightly_smiling_face: .

Reply:
I pushed another commit where I moved the checking/parsing of cookie value and the config file to before the cache_key is built. I then reuse that parsed data later on if/when the aggregate function uses them to create the list of EngineHandlers for the search.

Reply:
> I pushed another commit where I moved the checking/parsing of cookie value and the config file to before the cache_key is built. I then reuse that parsed data later on if/when the aggregate function uses them to create the list of EngineHandlers for the search.

Sorry, for the late reply.

I think you still didn't get my point :smile: . I think now I will provide you with a sample code which might give you a better idea on what I am suggesting. So I would suggest taking a look. :slightly_smiling_face: 

**Sample Code**
```diff
/// Fetches the results for a query and page. It First checks the redis cache, if that
/// fails it gets proper results by requesting from the upstream search engines.
///
/// # Arguments
///
/// * `url` - It takes the url of the current page that requested the search results for a
/// particular search query.
/// * `config` - It takes a parsed config struct.
/// * `query` - It takes the page number as u32 value.
/// * `req` - It takes the `HttpRequest` struct as a value.
///
/// # Error
///
/// It returns the `SearchResults` struct if the search results could be successfully fetched from
/// the cache or from the upstream search engines otherwise it returns an appropriate error.
async fn results(
    config: &Config,
    cache: &web::Data<SharedCache>,
    query: &str,
    page: u32,
    req: HttpRequest,
    safe_search: &Option<u8>,
) -> Result<SearchResults, Box<dyn std::error::Error>> {
    // eagerly parse cookie value to evaluate safe search level
    let cookie_value = req.cookie("appCookie");

    let cookie_value: Option<Cookie<'_>> = cookie_value
        .as_ref()
        .and_then(|cv| serde_json::from_str(cv.name_value().1).ok());

    let safe_search_level = get_safesearch_level(
        safe_search,
        &cookie_value.as_ref().map(|cv| cv.safe_search_level),
        config.safe_search,
    );

+   let mut cache_key = format!(
        "http://{}:{}/search?q={}&page={}&safesearch={}",
        config.binding_ip, config.port, query, page, safe_search_level
    );

    // fetch the cached results json.
    let cached_results = cache.cached_results(&cache_key).await;
    // check if fetched cache results was indeed fetched or it was an error and if so
    // handle the data accordingly.
    match cached_results {
        Ok(results) => Ok(results),
        Err(_) => {
            if safe_search_level == 4 {
                let mut results: SearchResults = SearchResults::default();

                let flag: bool =
                    !is_match_from_filter_list(file_path(FileType::BlockList)?, query)?;
                // Return early when query contains disallowed words,
                if flag {
                    results.set_disallowed();
                    cache.cache_results(&results, &cache_key).await?;
                    results.set_safe_search_level(safe_search_level);
                    return Ok(results);
                }
            }

            // check if the cookie value is empty or not if it is empty then use the
            // default selected upstream search engines from the config file otherwise
            // parse the non-empty cookie and grab the user selected engines from the
            // UI and use that.
            let mut results: SearchResults = match cookie_value {
                Some(cookie_value) => {
                    let engines: Vec<EngineHandler> = cookie_value
                        .engines
                        .iter()
                        .filter_map(|name| EngineHandler::new(name).ok())
                        .collect();

+                   let mut engine_temp = cookie_value.engines;
+                   engine_temp.sort();

+                   cache_key = format!("{cache_key}&engines={}", engine_temp.join(","));

+                   drop(engine_temp);

                    match engines.is_empty() {
                        false => {
                            aggregate(
                                query,
                                page,
                                config.aggregator.random_delay,
                                config.debug,
                                &engines,
                                config.request_timeout,
                                safe_search_level,
                            )
                            .await?
                        }
                        true => {
                            let mut search_results = SearchResults::default();
                            search_results.set_no_engines_selected();
                            search_results
                        }
                    }
                }
+               None => {
+                   cache_key = format!(
+                       "{cache_key}&engines={}",
+                       &config
+                           .upstream_search_engines
+                           .clone()
+                           .into_keys()
+                           .collect::<Vec<String>>()
+                           .join(",")
+                   );

+                   aggregate(
+                       query,
+                       page,
+                       config.aggregator.random_delay,
+                       config.debug,
+                       &config
+                           .upstream_search_engines
+                           .clone()
+                           .into_iter()
+                           .filter_map(|(key, value)| value.then_some(key))
+                           .map(|engine| EngineHandler::new(&engine))
+                           .collect::<Result<Vec<EngineHandler>, error_stack::Report<EngineError>>>(
+                           )?,
+                       config.request_timeout,
+                       safe_search_level,
+                   )
+                   .await?
+               }
+           };
            if results.engine_errors_info().is_empty()
                && results.results().is_empty()
                && !results.no_engines_selected()
            {
                results.set_filtered();
            }
            cache.cache_results(&results, &cache_key).await?;
            results.set_safe_search_level(safe_search_level);
            Ok(results)
        }
    }
} 
```

Reply:
This example would have a bug however, the cache_key variable is used almost immediately to get the cache_results,  therefore, if we try to retrieve results that don't have the modifications you want to do later on it wont return the same result. That's why we need to do all the modifications to cache_key and have its final value before 
`let cached_results = cache.cached_results(&cache_key).await;`

Reply:
Basically, in this example code, the key would be saved with the engine data, but wouldn't be loaded/checked with a key that has the engine data. So we can never read any of the data that's saved to the cache.

Reply:
> This example would have a bug however, the cache_key variable is used almost immediately to get the cache_results, therefore, if we try to retrieve results that don't have the modifications you want to do later on it wont return the same result. That's why we need to do all the modifications to cache_key and have its final value before `let cached_results = cache.cached_results(&cache_key).await;`

Yes, you are right, I almost forgot about that error, but now I have made several updates to the sample code which I have attached down below. So I would suggest having a look:

**src/modals/server_modals.rs**
```diff
 //! This module provides the models to parse cookies and search parameters from the search
 //! engine website.
 use serde::Deserialize;
 
+use super::parser_models::Style;
+
 /// A named struct which deserializes all the user provided search parameters and stores them.
 #[derive(Deserialize)]
 pub struct SearchParams {
     /// It stores the search parameter option `q` (or query in simple words)
     /// of the search url.
     pub q: Option<String>,
     /// It stores the search parameter `page` (or pageno in simple words)
     /// of the search url.
     pub page: Option<u32>,
     /// It stores the search parameter `safesearch` (or safe search level in simple words) of the
     /// search url.
     pub safesearch: Option<u8>,
 }
 
 /// A named struct which is used to deserialize the cookies fetched from the client side.
 #[allow(dead_code)]
 #[derive(Deserialize)]
 pub struct Cookie<'a> {
     /// It stores the theme name used in the website.
     pub theme: &'a str,
     /// It stores the colorscheme name used for the website theme.
     pub colorscheme: &'a str,
     /// It stores the user selected upstream search engines selected from the UI.
-    pub engines: Vec<&'a str>,
+    pub engines: Vec<String>,
     /// It stores the user selected safe search level from the UI.
     pub safe_search_level: u8,
 }
+
+impl<'a> Cookie<'a> {
+    ///
+    pub fn new(style: &'a Style, safe_search_level: u8, mut engines: Vec<String>) -> Self {
+        //
+        engines.sort();
+
+        Self {
+            theme: &style.theme,
+            colorscheme: &style.colorscheme,
+            engines,
+            safe_search_level,
+        }
+    }
+}
```

**src/server/routes/search.rs**
```diff
//! This module handles the search route of the search engine website.
 
 use crate::{
     cache::cacher::SharedCache,
     config::parser::Config,
     handler::{file_path, FileType},
     models::{
         aggregation_models::SearchResults,
         engine_models::{EngineError, EngineHandler},
         server_models::{Cookie, SearchParams},
     },
     results::aggregator::aggregate,
 };
 use actix_web::{get, http::header::ContentType, web, HttpRequest, HttpResponse};
 use regex::Regex;
 use std::{
     fs::File,
     io::{BufRead, BufReader, Read},
 };
 use tokio::join;
 
 /// Handles the route of search page of the `websurfx` meta search engine website and it takes
 /// two search url parameters `q` and `page` where `page` parameter is optional.
 ///
 /// # Example
 ///
 /// ```bash
 /// curl "http://127.0.0.1:8080/search?q=sweden&page=1"
 /// ```
 ///
 /// Or
 ///
 /// ```bash
 /// curl "http://127.0.0.1:8080/search?q=sweden"
 /// ```
 #[get("/search")]
 pub async fn search(
     req: HttpRequest,
     config: web::Data<Config>,
     cache: web::Data<SharedCache>,
 ) -> Result<HttpResponse, Box<dyn std::error::Error>> {
     let params = web::Query::<SearchParams>::from_query(req.query_string())?;
     match &params.q {
         Some(query) => {
             if query.trim().is_empty() {
                 return Ok(HttpResponse::TemporaryRedirect()
                     .insert_header(("location", "/"))
                     .finish());
             }
 
+            // eagerly parse cookie value to evaluate safe search level
+            let cookie_value = req.cookie("appCookie");
+
+            let cookie_value_or_config_values: Cookie<'_> = match cookie_value {
+                Some(ref cookie) => serde_json::from_str(cookie.value())?,
+                None => Cookie::new(
+                    &config.style,
+                    config.safe_search,
+                    config
+                        .upstream_search_engines
+                        .clone()
+                        .into_iter()
+                        .filter_map(|(key, value)| value.then_some(key))
+                        .collect(),
+                ),
+            };
+
+            let safe_search_level = get_safesearch_level(
+                params.safesearch.as_ref(),
+                cookie_value_or_config_values.safe_search_level,
+                config.safe_search,
+            );
+
             let get_results = |page| {
                 results(
                     &config,
                     &cache,
                     query,
                     page,
-                    req.clone(),
-                    &params.safesearch,
+                    &cookie_value_or_config_values,
+                    format!(
+                        "http://{}:{}/search?q={}&page={}&safesearch={}&engines={}",
+                        config.binding_ip,
+                        config.port,
+                        query,
+                        page,
+                        safe_search_level,
+                        cookie_value_or_config_values.engines.join(",")
+                    ),
                 )
             };
 
             // .max(1) makes sure that the page >= 0.
             let page = params.page.unwrap_or(1).max(1) - 1;
 
             let (_, results, _) = join!(
                 get_results(page.saturating_sub(1)),
                 get_results(page),
                 get_results(page + 1)
             );
 
             Ok(HttpResponse::Ok().content_type(ContentType::html()).body(
                 crate::templates::views::search::search(
                     &config.style.colorscheme,
                     &config.style.theme,
                     &config.style.animation,
                     query,
                     &results?,
                 )
                 .0,
             ))
         }
         None => Ok(HttpResponse::TemporaryRedirect()
             .insert_header(("location", "/"))
             .finish()),
     }
 }
 
 /// Fetches the results for a query and page. It First checks the redis cache, if that
 /// fails it gets proper results by requesting from the upstream search engines.
 ///
 /// # Arguments
 ///
 /// * `url` - It takes the url of the current page that requested the search results for a
 /// particular search query.
 /// * `config` - It takes a parsed config struct.
 /// * `query` - It takes the page number as u32 value.
 /// * `req` - It takes the `HttpRequest` struct as a value.
 ///
 /// # Error
 ///
 /// It returns the `SearchResults` struct if the search results could be successfully fetched from
 /// the cache or from the upstream search engines otherwise it returns an appropriate error.
 async fn results(
     config: &Config,
     cache: &web::Data<SharedCache>,
     query: &str,
     page: u32,
-    req: HttpRequest,
-    safe_search: &Option<u8>,
+    cookie_or_config: &Cookie<'_>,
+    cache_key: String,
 ) -> Result<SearchResults, Box<dyn std::error::Error>> {
-    // eagerly parse cookie value to evaluate safe search level
-    let cookie_value = req.cookie("appCookie");
-
-    let cookie_value: Option<Cookie<'_>> = cookie_value
-        .as_ref()
-        .and_then(|cv| serde_json::from_str(cv.name_value().1).ok());
-
-    let safe_search_level = get_safesearch_level(
-        safe_search,
-        &cookie_value.as_ref().map(|cv| cv.safe_search_level),
-        config.safe_search,
-    );
-
-    let cache_key = format!(
-        "http://{}:{}/search?q={}&page={}&safesearch={}",
-        config.binding_ip, config.port, query, page, safe_search_level
-    );
-
     // fetch the cached results json.
     let cached_results = cache.cached_results(&cache_key).await;
     // check if fetched cache results was indeed fetched or it was an error and if so
     // handle the data accordingly.
     match cached_results {
         Ok(results) => Ok(results),
         Err(_) => {
-            if safe_search_level == 4 {
+            if cookie_or_config.safe_search_level == 4 {
                 let mut results: SearchResults = SearchResults::default();
 
                 let flag: bool =
                     !is_match_from_filter_list(file_path(FileType::BlockList)?, query)?;
                 // Return early when query contains disallowed words,
                 if flag {
                     results.set_disallowed();
                     cache.cache_results(&results, &cache_key).await?;
-                    results.set_safe_search_level(safe_search_level);
+                    results.set_safe_search_level(cookie_or_config.safe_search_level);
                     return Ok(results);
                 }
             }
 
             // check if the cookie value is empty or not if it is empty then use the
             // default selected upstream search engines from the config file otherwise
             // parse the non-empty cookie and grab the user selected engines from the
             // UI and use that.
-            let mut results: SearchResults = match cookie_value {
-                Some(cookie_value) => {
-                    let engines: Vec<EngineHandler> = cookie_value
-                        .engines
-                        .iter()
-                        .filter_map(|name| EngineHandler::new(name).ok())
-                        .collect();
-
-                    match engines.is_empty() {
-                        false => {
-                            aggregate(
-                                query,
-                                page,
-                                config.aggregator.random_delay,
-                                config.debug,
-                                &engines,
-                                config.request_timeout,
-                                safe_search_level,
-                            )
-                            .await?
-                        }
-                        true => {
-                            let mut search_results = SearchResults::default();
-                            search_results.set_no_engines_selected();
-                            search_results
-                        }
-                    }
+            let engines: Vec<EngineHandler> = cookie_or_config
+                .engines
+                .iter()
+                .map(|engine| EngineHandler::new(engine))
+                .collect::<Result<Vec<EngineHandler>, error_stack::Report<EngineError>>>()?;
+
+            let mut results = match engines.is_empty() {
+                false => {
+                    aggregate(
+                        query,
+                        page,
+                        config.aggregator.random_delay,
+                        config.debug,
+                        &engines,
+                        config.request_timeout,
+                        cookie_or_config.safe_search_level,
+                    )
+                    .await?
+                }
+                true => {
+                    let mut search_results = SearchResults::default();
+                    search_results.set_no_engines_selected();
+                    search_results
                 }
-                None => aggregate(
-                    query,
-                    page,
-                    config.aggregator.random_delay,
-                    config.debug,
-                    &config
-                        .upstream_search_engines
-                        .clone()
-                        .into_iter()
-                        .filter_map(|(key, value)| value.then_some(key))
-                        .map(|engine| EngineHandler::new(&engine))
-                        .collect::<Result<Vec<EngineHandler>, error_stack::Report<EngineError>>>(
-                        )?,
-                    config.request_timeout,
-                    safe_search_level,
-                )
-                .await?,
             };
             if results.engine_errors_info().is_empty()
                 && results.results().is_empty()
                 && !results.no_engines_selected()
             {
                 results.set_filtered();
             }
             cache.cache_results(&results, &cache_key).await?;
-            results.set_safe_search_level(safe_search_level);
+            results.set_safe_search_level(cookie_or_config.safe_search_level);
             Ok(results)
         }
     }
 }
 
 /// A helper function which checks whether the search query contains any keywords which should be
 /// disallowed/allowed based on the regex based rules present in the blocklist and allowlist files.
 ///
 /// # Arguments
 ///
 /// * `file_path` - It takes the file path of the list as the argument.
 /// * `query` - It takes the search query to be checked against the list as an argument.
 ///
 /// # Error
 ///
 /// Returns a bool indicating whether the results were found in the list or not on success
 /// otherwise returns a standard error type on a failure.
 fn is_match_from_filter_list(
     file_path: &str,
     query: &str,
 ) -> Result<bool, Box<dyn std::error::Error>> {
     let mut reader = BufReader::new(File::open(file_path)?);
     for line in reader.by_ref().lines() {
         let re = Regex::new(&line?)?;
         if re.is_match(query) {
             return Ok(true);
         }
     }
 
     Ok(false)
 }
 
 /// A helper function which returns the safe search level based on the url params
 /// and cookie value.
 ///
 /// # Argurments
 ///
 /// * `safe_search` - Safe search level from the url.
 /// * `cookie` - User's cookie
 /// * `default` - Safe search level to fall back to
-fn get_safesearch_level(safe_search: &Option<u8>, cookie: &Option<u8>, default: u8) -> u8 {
-    match safe_search {
+fn get_safesearch_level(
+    config_safe_search: Option<&u8>,
+    cookie_safe_search_level: u8,
+    default: u8,
+) -> u8 {
+    match config_safe_search {
         Some(ss) => {
             if *ss >= 3 {
                 default
             } else {
                 *ss
             }
         }
-        None => cookie.unwrap_or(default),
+        None => cookie_safe_search_level,
     }
 }
```

Also, note the sample code that I have attached this time affects code from multiple files which include `search.rs` and `server_modals` file. Additionally, I have not added/updated comments/documentation in some area because it is a sample code and may not be perfect and may require refactoring or changing the code in some area. So I would suggest adding/updating comments/documentation to the code if necesssary from your side. :slightly_smiling_face:

Reply:
> Yeah this should work, I was trying to make logic changes without changing any function parameters or definitions. What if we, in an attempt to make it a bit more rusty, create a sort of result argument struct that has a default implementation using the config file and can be changed with the engines found in the cookie. We can then pass the struct as a singe parameter for the result function to use.
> 
> We might need to make a change like that later to make sure everything the result function requests is Send and 'static or otherwise works with tokio::spawn.

Yes, that would be better, I think we can have those changes in another PR after this. What do you say? :slightly_smiling_face: .